### ISSUE TITLE
 Is there a way to enforce receiving a client certificate?

### ISSUE THREAD

**Issue by rab5nu3** [randombit/botan#1937]
Hi, I'm using Botan 2.10.0 and have a TLS server that requires client authentication. Thus I returned a store in trusted_certificate_authorities() that contains the corresponding CAs. If the client now doesn't send a client certificate, the following code in process_client_key_exchange_msg() leads to just skipping the verification step and silently jumps to the next state HANDSHAKE_CCS:

```
if(pending_state.received_handshake_msg(CERTIFICATE) && !pending_state.client_certs()->empty())
      pending_state.set_expected_next(CERTIFICATE_VERIFY);
   else
      pending_state.set_expected_next(HANDSHAKE_CCS);
```

Is there a way to enforce receiving a client certificate in this situation?

**Comment by randombit** [comment]
There should be some way to interrupt the handshake, I think you are right that this is missing. After the handshake completes you can examine return value of `Channel::peer_cert_chain` and send an alert if it is empty.

**Comment by randombit** [comment]
Oh you could also add a similar check in the callback  `tls_session_established` at that point the peer cert chain is known and stored as part of the passed Session object.

**Comment by rab5nu3** [comment]
Thank you for your answer. Just to clarify: Is your proposal a workaround or will there be a fix? Checking in  tls_session_established seems to me a bit circuitous compared to other API functionality :-). If there won't be a fix, could we also check in tls_inspect_handshake_msg, e.g. in state change_cypher_spec?

Just to complete my original question: There are two cases, the client doesn't send a certificate at all or it sends an empty certificate. Currently, this doesn't make a difference.

**Comment by randombit** [comment]
Checking in the inspection callback certainly works. The actual TLS handshake message classes in `tls_messages.h` are not really considered part of the stable API, sometimes supporting new protocol features does require modifying them, but they are available and I don't see any reason that `Certificate::cert_chain` would change much in the future, which is all you would need to look at in this case.

I do think there should be a cleaner way to express this since it's a common need and telling everyone to just use a not-stable API isn't so hot. Probably the simplest is adding a new Policy toggle `client_cert_authentication_required` which if true causes the server to send a CertificateRequest even if the client CA list is empty, and which additionally requires the client send a non-empty Certificate message.

> There are two cases, (1.) the client doesn't send a certificate at all or (2.) it sends an empty certificate. Currently, this doesn't make a difference when handled in Botan code.

That doesn't seem quite right. SSLv3 (which is no longer supported) allowed the client to omit sending the Certificate message but all later versions require an empty message be sent if a CertificateRequest was given. We just don't expect a CertificateVerify if the Certificate message was empty.

**Comment by rab5nu3** [comment]
> That doesn't seem quite right. SSLv3 (which is no longer supported) allowed the client to omit sending the Certificate message but all later versions require an empty message be sent if a CertificateRequest was given.

But shouldn't then an alert message be sent in the case that the certificate message is missing? RFC 8446 defines the alert type certificate_required for that case, doesn't it?

And receiving an empty list of client certificates in fact means that the client is unauthorized. Is this somewhere reflected via the Botan API or is it completely up to the user to handle that case in tls_inspect_handshake_msg?

**Comment by randombit** [comment]
@rab5nu3 Can you try the patch in #1942 and see if it meets your needs? It should apply cleanly to 2.10.0 release.

**Comment by rab5nu3** [comment]
> Not necessarily. It is allowed and sometimes useful to ask for a cert but allow the handshake to continue even if none is presented. Eg, the application might allow login via certs, but if it is not available then the user is presented with a password-based login form. In such a case, you would not want to abort the handshake.

Not sure if you understood me correctly. I didn't meant the case where an empty certificate message is sent as an answer from the client to the server. I meant the case the client omits the whole message which isn't allowed according to the RFC as you've already mentioned.

Thank you for the patch. I'll give it a try this afternoon, hopefully.

**Comment by randombit** [comment]
> Not sure if you understood me correctly. I didn't meant the case where an empty certificate message is sent as an answer from the client to the server. I meant the case the client omits the whole message which isn't allowed according to the RFC as you've already mentioned.

Wait, have you observed this behavior? https://github.com/randombit/botan/blob/master/src/lib/tls/tls_server.cpp#L874 requires a cert message be sent by the client if a certificate request is made, and if that is not working it suggests there is a serious problem with the handshake validation.

**Comment by randombit** [comment]
> I meant the case the client omits the whole message

I just tried this with a hacked client, server rejects. "Unexpected state transition in handshake, got type 16 expected certificate received client_hello,client_key_exchange"

**Comment by rab5nu3** [comment]
Sorry for my late reply... I finally managed to check the patch from #1942 today. It works quite well but I'm a bit surprised that you throw a CERTIFICATE_REQUIRED alert since you rightly mentioned that this is TLS 1.3 specific. What do you think about this?

**Comment by randombit** [comment]
In the version of that PR that got merged (#1954) we use handshake_failure because that is what the BoringSSL test suite expects for this situation in versions < 1.3
